### ISSUE TITLE
lxc-attach breakage on kernel 5.12.9 and higher

### ISSUE THREAD

**Issue by msizanoen1** [lxc/lxc#3863]
# Required information

 * Distribution: Fedora
 * Distribution version: 34
 * The output of
   * `lxc-start --version`: 4.0.0-devel
   * `lxc-checkconfig`
```
LXC version 4.0.0-devel
Kernel configuration not found at /proc/config.gz; searching...
Kernel configuration found at /boot/config-5.12.9-300.fc34.x86_64
--- Namespaces ---
Namespaces: enabled
Utsname namespace: enabled
Ipc namespace: enabled
Pid namespace: enabled
User namespace: enabled
Warning: newuidmap is not setuid-root
Warning: newgidmap is not setuid-root
Network namespace: enabled

--- Control groups ---
Cgroups: enabled

Cgroup v1 mount points: 


Cgroup v2 mount points: 
/sys/fs/cgroup

Cgroup v1 systemd controller: missing
Cgroup v1 freezer controller: missing
Cgroup namespace: required
Cgroup device: enabled
Cgroup sched: enabled
Cgroup cpu account: enabled
Cgroup memory controller: enabled
Cgroup cpuset: enabled

--- Misc ---
Veth pair device: enabled, loaded
Macvlan: enabled, not loaded
Vlan: enabled, not loaded
Bridges: enabled, loaded
Advanced netfilter: enabled, loaded
CONFIG_NF_NAT_IPV4: missing
CONFIG_NF_NAT_IPV6: missing
CONFIG_IP_NF_TARGET_MASQUERADE: enabled, not loaded
CONFIG_IP6_NF_TARGET_MASQUERADE: enabled, not loaded
CONFIG_NETFILTER_XT_TARGET_CHECKSUM: enabled, loaded
CONFIG_NETFILTER_XT_MATCH_COMMENT: enabled, not loaded
FUSE (for use with lxcfs): enabled, loaded

--- Checkpoint/Restore ---
checkpoint restore: enabled
CONFIG_FHANDLE: enabled
CONFIG_EVENTFD: enabled
CONFIG_EPOLL: enabled
CONFIG_UNIX_DIAG: enabled
CONFIG_INET_DIAG: enabled
CONFIG_PACKET_DIAG: enabled
CONFIG_NETLINK_DIAG: enabled
File capabilities: 

Note : Before booting a new kernel, you can check its configuration
usage : CONFIG=/path/to/config /usr/local/bin/lxc-checkconfig
```
   * `uname -a`: `Linux [hostname redacted] 5.12.9-300.fc34.x86_64 #1 SMP Thu Jun 3 13:51:40 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`
   * `cat /proc/self/cgroup`: N/A
   * `cat /proc/1/mounts`: N/A

# Issue description

Linux kernel commit [bfb819ea20ce8bbeeba17e1a6418bf8bda91fc28](https://github.com/torvalds/linux/commit/bfb819ea20ce8bbeeba17e1a6418bf8bda91fc28) which has since been backported to stable kernel 5.2.19 causes breakage in lxc-attach.
This is caused because lxc-attach [opens a file descriptor for `/proc/<pid>/attr/...` before changing credentials](https://github.com/lxc/lxc/blob/a197d2fb5bcbbc722cebb3fb9c3add685bf94cd2/src/lxc/attach.c#L1184-L1195).

Proposed fix:
Opening the `/proc/<pid>/attr/...` file before changing credentials shouldn't be necessary because the permissions for those files are all 666, instead the actual security checks are done at `write()` time:
```
$ ls -l /proc/1/attr
total 0
-rw-rw-rw-. 1 root root 0 Jun  9 21:37 current
-rw-rw-rw-. 1 root root 0 Jun  9 21:37 exec
-rw-rw-rw-. 1 root root 0 Jun  9 21:37 fscreate
-rw-rw-rw-. 1 root root 0 Jun  9 21:37 keycreate
-r--r--r--. 1 root root 0 Jun  9 21:37 prev
-rw-rw-rw-. 1 root root 0 Jun  9 21:37 sockcreate
```
Therefore to fix this instead of opening `/proc/<pid>/attr/...` before initiating credential changes lxc-attach can simply open `/proc/self/attr/...` and write the new security context there.

Note regarding hidepid: The process should always be able to access it's own /proc directories as shown in the kernel source: procfs use the `has_pid_permissions` function to check whether the current process has access to another processes' /proc directory, which calls the `ptrace_may_access` function which [states that a process may always access any /proc directories of threads belonging to the process](https://github.com/torvalds/linux/blob/dbb5afad100a828c97e012c6106566d99f041db6/kernel/ptrace.c#L314-L315).
# Steps to reproduce

 1. Step one: Make sure the kernel is 5.12.9 or higher or have the above commit, and some form of Mandatory Access Control is enabled
 2. Step two: Start a container
 3. Step three: Try to lxc-attach the container

**Comment by simondeziel** [comment]
FYI, @brauner is discussing this upstream, see https://lore.kernel.org/regressions/20210607142245.eikvyeacqwwu6dn3@wittgenstein/

**Comment by brauner** [comment]
Irregardless that we could do this differently this is a regression. Kees has proposed a fix for this.

**Comment by brauner** [comment]
Also note that there are scenarios where we can't do this. We don't rely on the container's procfs instance ever and we might not even have a procfs instance mounted so it should really be the parent who does this.

**Comment by msizanoen1** [comment]
That scenario can be worked around by opening the `/proc/<pid>` handle in the parent and passing that instead of the LSM label FD and open the actual LSM label file in the child process.

**Comment by brauner** [comment]
> That scenario can be worked around by opening the `/proc/<pid>` handle in the parent and passing that instead of the LSM label FD and open the actual LSM label file in the child process.

Nope, we can't do that because that would allow the container to attack the host. We used to do that and switched away from that. To quote:
> We opened a file descriptor refering to proc on the host inside the host namespace and handed that fd to the attached process in attach_child_main(). This was done to ensure that LSM labels were correctly setup. However, by exploiting a potential kernel bug, ptrace could be used to prevent the file descriptor from being closed which in turn could be used by an  unprivileged container to gain access to the host namespace. Aside from this needing an upstream kernel fix, we should make sure that we don't pass the fd for proc itself to the attached process. [...]
To minimize the attack surface, we only send the fd for /proc/self/attr/exec or /proc/self/attr/current to the attached process. To do this we introduce a little more IPC between the child and parent [...]

**Comment by brauner** [comment]
Or in other words, we used to send around /proc itself and we could send /proc/<pid> but I'm not a fan of sending around dirfds when we don't need to. We've been bitten by bugs before.

**Comment by brauner** [comment]
Fix for this in the kernel is merged btw
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=591a22c14d3f45cc38bd1931c593c221df2f1881

**Comment by msizanoen1** [comment]
This issue can be closed when the fix got backported to stable and longterm kernels.

**Comment by msizanoen1** [comment]
Fixed in Linux kernel v5.12.11, issue can be closed.
